File: benchmarks/fixtures/webapp_ts/src/cache/memoryCache.ts
/** In-memory LRU cache. */
import { getLogger } from '../utils/helpers';
import { BaseCache } from './base';

const logger = getLogger("cache.memory");

/** LRU memory cache. */
export class MemoryCache extends BaseCache {
    private store: Map<string, unknown> = new Map();
    private expiry: Map<string, number> = new Map();
    private maxSize: number;

    constructor(maxSize: number = 1000) {
        super("memory");
        this.maxSize = maxSize;
        logger.info(`MemoryCache created: maxSize=${maxSize}`);
    }

    get(key: string): unknown | null {
        if (this.store.has(key)) {
            const exp = this.expiry.get(key) ?? Infinity;
            if (Date.now() > exp) {
                this.store.delete(key);
                this.expiry.delete(key);
                this.misses++;
                return null;
            }
            this.hits++;
            // Move to end for LRU
            const val = this.store.get(key);
            this.store.delete(key);
            this.store.set(key, val);
            return val ?? null;
        }
        this.misses++;
        return null;
    }

    set(key: string, value: unknown, ttl: number = 300): void {
        if (this.store.has(key)) {
            this.store.delete(key);
        } else if (this.store.size >= this.maxSize) {
            const firstKey = this.store.keys().next().value;
            if (firstKey !== undefined) {
                this.store.delete(firstKey);
                this.expiry.delete(firstKey);
                logger.info(`LRU evicted: ${firstKey}`);
            }
        }
        this.store.set(key, value);
        this.expiry.set(key, Date.now() + ttl * 1000);
    }

    delete(key: string): boolean {
        this.expiry.delete(key);
        return this.store.delete(key);
    }

    clear(): number {
        const count = this.store.size;
        this.store.clear();
        this.expiry.clear();
        return count;
    }

    size(): number {
        return this.store.size;
    }
}

File: benchmarks/fixtures/webapp_rb/cache/memory_cache.rb
# In-memory LRU cache.

require_relative '../utils/helpers'
require_relative 'base_cache'

# LRU memory cache.
class MemoryCache < BaseCache
  def initialize(max_size = 1000)
    super('memory')
    @store = {}
    @expiry = {}
    @max_size = max_size
    get_logger('cache.memory').info("MemoryCache created: max_size=#{max_size}")
  end

  def get(key)
    if @store.key?(key)
      exp = @expiry[key] || Float::INFINITY
      if Time.now.to_i > exp
        @store.delete(key)
        @expiry.delete(key)
        @misses += 1
        return nil
      end
      @hits += 1
      # Move to end for LRU
      val = @store.delete(key)
      @store[key] = val
      return val
    end
    @misses += 1
    nil
  end

  def set(key, value, ttl = 300)
    if @store.key?(key)
      @store.delete(key)
    elsif @store.size >= @max_size
      first_key = @store.keys.first
      if first_key
        @store.delete(first_key)
        @expiry.delete(first_key)
        get_logger('cache.memory').info("LRU evicted: #{first_key}")
      end
    end
    @store[key] = value
    @expiry[key] = Time.now.to_i + ttl
  end

  def delete(key)
    @expiry.delete(key)
    !@store.delete(key).nil?
  end

  def clear
    count = @store.size
    @store.clear
    @expiry.clear
    count
  end

  def size
    @store.size
  end
end
